torchtune/models/qwen2/_tokenizer.py

```diff
@@ -356,68 +416,39 @@ def tokenize_messages(
             else messages
         )
 
-        tokenized_messages = []
+        tokens = []
         mask = []
-        for index, message in enumerate(templated_messages):
-            tokens = []
-
-            # message header
-            if message.role != "ipython":
-                tokens.append(self.im_start_id)
-                tokens.extend(
-                    self.encode(f"{message.role}\n", add_bos=False, add_eos=False)
-                )
 
-            # message content
-            for item in message.content:
-                if item["type"] == "text":
-                    tokens.extend(
-                        self.encode(
-                            item["content"],
-                            add_bos=False,
-                            add_eos=False,
-                        )
-                    )
-                else:
-                    raise RuntimeError(
-                        f"Unsupported message content type: {item['type']}"
-                    )
-
-            # message footer
-            if message.role != "ipython" and (
-                message.role != "assistant" or index != len(messages) - 1
-            ):
-                tokens.append(self.im_end_id)
-                tokens.extend(self.encode("\n", add_bos=False, add_eos=False))
-
-            tokenized_messages.extend(tokens)
-            mask.extend([message.masked] * len(tokens))
-
-            # Break out early if we reach max_seq_len
-            if self.max_seq_len and len(tokenized_messages) >= self.max_seq_len:
+        num_messages = len(templated_messages)
+        for i, message in enumerate(templated_messages):
+            tokenized_message = self.tokenize_message(
+                message, index=i, num_messages=num_messages
+            )
+            tokens = tokens + tokenized_message
+            mask = mask + ([message.masked] * len(tokenized_message))
+
+            if self.max_seq_len and len(tokens) >= self.max_seq_len:
                 break
 
-        # Add the End-Of-Sequence token
-        if add_eos:
-            tokenized_messages.append(self.eos_id)
-            mask.append(mask[-1])
+        if add_end_tokens:
+            tokens = tokens + [self.eos_id]
+            mask = mask + [mask[-1] if mask else True]
```

Comment:
Noob q: why is this different than what we're doing for Llama3? [ref](https://github.com/pytorch/torchtune/blob/c3c532a9e74a835c67da94bb1ac64a119ab59b8f/torchtune/models/llama3/_tokenizer.py#L328)